_posts/2018-05-07-is-graphql-the-future.markdown

```diff
@@ -0,0 +1,188 @@
+
+Often, we just resolve an object field to a domain object. Its scalar fields might correspond to properties of that domain object and its object fields might correspond to related objects. But the architecture is more powerful than this! A deeply nested field can potentially be the result of the resolved values of all its parents. It all depends on how you design your resolvers to work together.
+
+This pattern reminds me a bit of when [jQuery](https://api.jquery.com/) first clicked for me. A lot of details are propagated invisibly within your `jquery` object as you chain method calls to refine your DOM selections.
+
+## Looping and recursion
+
+GraphQL doesn’t have them, plain and simple. Consequently, the GraphQL DSLs you design are not [Turing-complete](https://en.wikipedia.org/wiki/Turing_completeness)--they will always halt in a finite amount of steps. This is really important, because it prevents clients from being able to send servers on errands that will never end. Of course, the _implementations_ of field resolvers on the server are free to do whatever they want in full Turing-complete glory.
+
+# The post-REST world
+
+Subtly, this paradigm is a sharp step away from a whole body of knowledge that models APIs as resources with fixed verbs, which we know as REST. It’s more appropriate to think of GraphQL requests as a script of remote procedure calls (RPC). From this perspective, the design of the schema is a lot less about data modeling than it is a question of how you want your entire API to be traversed. This encourages a verb-oriented mindset.
+
+## Verb orientation
+
+Speaking of verbs, you can think of _query_ as being the default verb in GraphQL. You model other verbs as mutations. I delayed learning about mutations, because I thought they must be way more complex than queries. Quite the opposite! They all sit in one big, flat bucket at the root of your schema, as the fields of the root `mutation` type. These fields have a type too, and if it is an object type, then you can issue effectively issue any number of follow-up queries after your mutation completes. Learning about mutations was when it really dawned on me that _fields are just function calls_.
+
+Mutations are a major break with REST. Separating them from the query structure means that they are not verbs on a resource, but verbs _on your entire service_. This eliminates one of REST’s key weak points, namely that complex operations that touch multiple parts of an application’s data model are difficult to model as a PUT, DELETE, POST, or PATCH on a single resource. In my experience, this "impedance mismatch” between API modeling and domain modeling has led to the worst aspects of my HTTP API designs.
```

Comment:
> but verbs _on your entire service_.

Yeah, agreed, I came to this same conclusion [quite recently in Peril](https://github.com/danger/peril/blob/22dd1137801442134dcdf336c86887602f4861fa/source/api/graphql/index.ts#L123-L132)